tedana/selection/selection_utils.py

```diff
@@ -313,6 +317,103 @@ def clean_dataframe(component_table):
 #################################################
 
 
+def expand_dict(node, field, metrics):
+    """Expand a dictionary with regular expressions.
+
+    Parameters
+    ----------
+    node : dict
+        A dictionary containing nested dictionaries called "parameters" and,
+        optionally, "kwargs".
+        Any of the values in the "parameters" or "kwargs" dictionaries may be
+        a regular expression, denoted by starting with "^".
+    field : :obj:`str`
+        The key in the ``node`` dictionary that should be expanded.
+    metrics : list of str
+        List of metric names to compare regular expressions against.
+
+    Returns
+    -------
+    regex_found : :obj:`bool`
+        True if any regular expressions were found in the input dictionary
+    out_nodes : :obj:`list[dict]`
+        A list of dictionaries with regular expressions replaced by all
+        matching values in the 'metrics' list.
+    """
+    regex_found = False
+    out_nodes = []
+    for k, v in node.get(field, {}).items():
+        if isinstance(v, str) and v.startswith("^"):
+            regex_found = True
+            replacements = [metric for metric in metrics if re.match(v, metric)]
+            if not replacements:
+                raise ValueError(f"No metrics matching regex '{v}' found.")
+
+            for replacement in replacements:
+                LGR.warning(f"Replacing {v} with {replacement}")
+                mod_node = copy.deepcopy(node)
+                mod_node[field][k] = replacement
+                out_nodes.append(mod_node)
+
+    return regex_found, out_nodes
+
+
+def expand_node(node, metrics):
+    """Expand node definitions with regular expressions.
+
+    Recursively expand a node so that any regular expressions are replaced with
+    any matching values in the 'metrics' list.
+    Regular expressions may be present as the value of any key in the
+    subdictionaries "parameters" or "kwargs".
+
+    Parameters
+    ----------
+    node : dict
+        A dictionary containing nested dictionaries called "parameters" and,
+        optionally, "kwargs".
+        Any of the values in the "parameters" or "kwargs" dictionaries may be
+        a regular expression, denoted by starting with "^".
+    metrics : list of str
+        List of metric names.
+    """
+    regex_found, out_nodes = expand_dict(node, "parameters", metrics)
+    if not regex_found:
+        regex_found, out_nodes = expand_dict(node, "kwargs", metrics)
+
+    if not regex_found:
+        # Stop early and just return the node if no regular expressions were found
+        out_nodes = [copy.deepcopy(node)]
+        return out_nodes
+
+    real_out_nodes = []
+    for out_node in out_nodes:
```

Comment:
@tsalo I'm working on improving documentation and type hints and I'm realizing I'm confused with part of what you did here. `expand_node` passes a single `node` dictionary to `expand_node`. `expand_node` takes two fields in this dictionary (parameters and kwargs) and sends them to `expand_dict` to identify and expand any regular expressions. After this is done, `expand_node` then recurses if any regular expressions were found. I'm having trouble seeing situations where this recursion would actually happen. I vaguely remember an earlier version where recursion did the expansion, but maybe this isn't needed anymore?

There also aren't tests for these two functions so I need to better understand when recursion should happen to write a test.

Reply:
The idea was to use the regular expression to find a list of columns, and then to run the correlation separately for each of them, but since the list of columns is fed into a single F-test now, I don't think this step is necessary anymore.